lib/Dialect/FIRRTL/Transforms/GrandCentralTaps.cpp

```diff
@@ -0,0 +1,683 @@
+        // TODO: This would be better handled through a proper cross-module
+        // reference preserved in the IR, such that ExportVerilog can insert a
+        // proper name here at the last moment.
+        auto name = op->getAttrOfType<StringAttr>("name");
+        if (!name) {
+          auto diag = op->emitError("target of memory tap must have a name");
+          diag.attachNote(blackBox.extModule->getLoc()) << llvm::formatv(
+              "used by MemTapAnnotation on port {0} here:", portName);
+          signalPassFailure();
+          continue;
+        }
+
+        // Formulate a hierarchical reference into the memory.
+        // CAVEAT: This just assumes that the memory will map to something that
+        // can be indexed in the final Verilog. If the memory gets turned into
+        // an instance of some sort, we lack the information necessary to go in
+        // and access individual elements of it. This will break horribly since
+        // we generate memory impls out-of-line already, and memories coming
+        // from an external generator are even worse. This needs a special node
+        // in the IR that can properly inject the memory array on emission.
+        wiring.prefices =
+            instanceGraph.getAbsolutePaths(op->getParentOfType<FModuleOp>());
+        wiring.suffix = llvm::formatv("{0}[{1}]", name.getValue(),
+                                      memPortIdx[portAnno.anno]++);
+        portWiring.push_back(std::move(wiring));
+        continue;
+      }
+
+      llvm_unreachable(
+          llvm::formatv("unknown annotation {}", portAnno.anno).str().c_str());
+    }
+
+    LLVM_DEBUG(llvm::dbgs() << "- Wire up as follows:\n");
+    for (auto wiring : portWiring) {
+      LLVM_DEBUG(llvm::dbgs() << "- Port " << wiring.portNum << ":\n");
+      for (auto path : wiring.prefices) {
+        LLVM_DEBUG(llvm::dbgs()
+                   << "  - " << path << "." << wiring.suffix << "\n");
+      }
+    }
+
+    // Now we have an awkward mapping problem. We have multiple data tap module
+    // instances, which reference things in modules that in turn have multiple
+    // instances. This is a side-effect of how Grand Central annotates things on
+    // modules rather than instances. (However in practice these will have a
+    // one-to-one correspondence due to CHIRRTL having fully uniquified
+    // instances.) To solve this issue, create a dedicated implementation for
+    // every data tap instance, and among the possible targets for the data taps
+    // choose the one with the shortest relative path to the data tap instance.
+    OpBuilder builder(blackBox.extModule);
+    unsigned implIdx = 0;
+    for (auto path : paths) {
+      builder.setInsertionPointAfter(blackBox.extModule);
+
+      // Create a new firrtl.module that implements the data tap.
+      auto name = StringAttr::get(
+          &getContext(),
+          llvm::formatv("{0}_impl_{1}", blackBox.extModule.getName(), implIdx++)
```

Comment:
Can this use Twine instead of format?  It's generally more efficient and is a bit simpler

Reply:
Good point, done!